From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Fri, 6 May 2022 15:29:09 -0700
Subject: [PATCH] Cache GameEvent lookups

This will build a cache of bukkit game events overtime as the lookup can be quite heavy with excessive vanilla game events being fired.

diff --git a/src/main/java/dev/yive/pluto/PlutoConfig.java b/src/main/java/dev/yive/pluto/PlutoConfig.java
index 6c64f94cbc9a436ee4326cd9e561a7a03c62e997..01f7fd798b01c783588db53d1618b8cd0b7b2978 100644
--- a/src/main/java/dev/yive/pluto/PlutoConfig.java
+++ b/src/main/java/dev/yive/pluto/PlutoConfig.java
@@ -194,5 +194,10 @@ public class PlutoConfig {
             this.worldDefaults.addDefault(path, Double.valueOf(dfl));
             return this.config.getDouble(path, this.worldDefaults.getDouble(path));
         }
+
+        public boolean useGameEventCache = true;
+        private void shouldUseGameEventCache() {
+            useGameEventCache = getBoolean("misc.use-game-events-cache", useGameEventCache);
+        }
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/gameevent/GameEventDispatcher.java b/src/main/java/net/minecraft/world/level/gameevent/GameEventDispatcher.java
index 37ae0f928440af5b2fdfe68ad6c9c54e8c95c82c..36fa4531886d776844c79464bddfb02146579f8c 100644
--- a/src/main/java/net/minecraft/world/level/gameevent/GameEventDispatcher.java
+++ b/src/main/java/net/minecraft/world/level/gameevent/GameEventDispatcher.java
@@ -25,17 +25,28 @@ public class GameEventDispatcher {
         this.level = world;
     }
 
+    private static final it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<String, org.bukkit.GameEvent> GAME_EVENT_CACHE = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>(); // Pluto - Cache GameEvent lookups
     public void post(GameEvent event, Vec3 emitterPos, GameEvent.Context emitter) {
         int i = event.getNotificationRadius();
         BlockPos blockposition = BlockPos.containing(emitterPos);
+        // Pluto start - Cache GameEvent lookups
+        org.bukkit.GameEvent bukkitEvent;
+        if (this.level.plutoConfig().useGameEventCache) {
+            bukkitEvent = GAME_EVENT_CACHE.computeIfAbsent(event.builtInRegistryHolder().key().location().getPath(), s -> CraftGameEvent.minecraftToBukkit(event));
+        } else {
+            bukkitEvent = CraftGameEvent.minecraftToBukkit(event);
+        }
+        if (bukkitEvent != null) {
         // CraftBukkit start
-        GenericGameEvent event1 = new GenericGameEvent(CraftGameEvent.minecraftToBukkit(event), CraftLocation.toBukkit(blockposition, this.level.getWorld()), (emitter.sourceEntity() == null) ? null : emitter.sourceEntity().getBukkitEntity(), i, !Bukkit.isPrimaryThread());
+        GenericGameEvent event1 = new GenericGameEvent(bukkitEvent, CraftLocation.toBukkit(blockposition, this.level.getWorld()), (emitter.sourceEntity() == null) ? null : emitter.sourceEntity().getBukkitEntity(), i, !Bukkit.isPrimaryThread());
+        // Pluto end - Cache GameEvent lookups
         this.level.getCraftServer().getPluginManager().callEvent(event1);
         if (event1.isCancelled()) {
             return;
         }
         i = event1.getRadius();
         // CraftBukkit end
+        } // Pluto - Cache GameEvent lookups
         int j = SectionPos.blockToSectionCoord(blockposition.getX() - i);
         int k = SectionPos.blockToSectionCoord(blockposition.getY() - i);
         int l = SectionPos.blockToSectionCoord(blockposition.getZ() - i);
